File: iOS/example/ai_companion_uikit/aiagent/voicecall/ZegoVoiceActivityChecker.m
#import "ZegoVoiceActivityChecker.h"

@implementation VadCheckerInfo
@end

@interface ZegoVoiceActivityChecker(){
    NSMutableArray<NSNumber*> *_queue;
    NSUInteger _capacity;
    float _threshold; //>=0,<=1,目前设置为10/15
    BOOL _voiceActivity; //YES表示说话状态，NO反之
    int64_t _checkSeq; //
}

@end

@implementation ZegoVoiceActivityChecker

- (instancetype)init {
    self = [super init];
    if (self) {
        _capacity = 7; //默认窗口500ms尺寸
        _queue = [[NSMutableArray alloc]initWithCapacity:_capacity];
        _voiceActivity = NO;
        _threshold = 23/28.0;
        _checkSeq = 0;
        for (int i=0; i<_capacity; i++) {
            [self enqueue:0];
        }
    }
    return self;
}

-(void)enqueue:(NSUInteger)vadValue{
    if ([self isFull]) {
        [self dequeue];
    }
    [_queue addObject:[NSNumber numberWithUnsignedInteger:vadValue]];
    
}

- (VadCheckerInfo*)voiceActivityDetection:(NSUInteger)vadValue {
    [self enqueue:vadValue];
    float weightAverage=0.0;
    for (int i= _queue.count; i>0; i--) {
        NSNumber* item = [_queue objectAtIndex:i-1];
        weightAverage += (i*[item intValue])/28.0;
    }
    if (weightAverage >= _threshold) {
        if (_voiceActivity == NO) {
            _checkSeq++;
        }
        _voiceActivity = YES;
    }else{
        _voiceActivity = NO;
    }
    
    VadCheckerInfo* chekerInfo = [[VadCheckerInfo alloc] init];
    chekerInfo.weightAverage = weightAverage;
    chekerInfo.voiceActivity = _voiceActivity;
    chekerInfo.checkSeq = _checkSeq;
    
    return chekerInfo;
}

- (id)dequeue {
    if (![self isEmpty]) {
        id object = _queue[0];
        [_queue removeObjectAtIndex:0];
        return object;
    } else {
        NSLog(@"Queue is empty. Cannot dequeue object.");
        return nil;
    }
}

- (BOOL)isEmpty {
    return [_queue count] == 0;
}

- (BOOL)isFull {
    return [_queue count] >= _capacity;
}
@end

File: android/ai_agent_android/aiagent/src/main/java/im/zego/aiagent/core/widget/ZegoVoiceActivityChecker.java
package im.zego.aiagent.core.widget;

import java.util.ArrayList;
import java.util.List;

/**
 * 快速打断相关类
 */
public class ZegoVoiceActivityChecker {

    private List<Integer> queue;
    private int capacity;
    private float threshold; // >=0, <=1, 目前设置为10/15
    private boolean voiceActivity; // true 表示说话状态，false 反之
    private long checkSeq;

    public ZegoVoiceActivityChecker() {
        this.capacity = 5; // 默认窗口500ms尺寸
        this.queue = new ArrayList<>(capacity);
        this.voiceActivity = false;
        this.threshold = 11.0f / 15.0f;
        this.checkSeq = 0;
        for (int i = 0; i < capacity; i++) {
            enqueue(0);
        }
    }

    public void enqueue(int vadValue) {
        if (isFull()) {
            dequeue();
        }
        queue.add(vadValue);
    }

    public VadCheckerInfo voiceActivityDetection(int vadValue) {
        enqueue(vadValue);
        float weightAverage = 0.0f;
        for (int i = queue.size(); i > 0; i--) {
            int item = queue.get(i - 1);
            weightAverage += (i * item) / 15.0f;
        }
        if (weightAverage >= threshold) {
            if (!voiceActivity) {
                checkSeq++;
            }
            voiceActivity = true;
        } else {
            voiceActivity = false;
        }

        VadCheckerInfo checkerInfo = new VadCheckerInfo();
        checkerInfo.setWeightAverage(weightAverage);
        checkerInfo.setVoiceActivity(voiceActivity);
        checkerInfo.setCheckSeq(checkSeq);

        return checkerInfo;
    }

    public Integer dequeue() {
        if (!isEmpty()) {
            Integer object = queue.remove(0);
            return object;
        } else {
            System.out.println("Queue is empty. Cannot dequeue object.");
            return null;
        }
    }

    public boolean isEmpty() {
        return queue.isEmpty();
    }

    public boolean isFull() {
        return queue.size() >= capacity;
    }

    public static class VadCheckerInfo {

        private float weightAverage;
        private boolean voiceActivity;
        private long checkSeq;

        public float getWeightAverage() {
            return weightAverage;
        }

        public void setWeightAverage(float weightAverage) {
            this.weightAverage = weightAverage;
        }

        public boolean isVoiceActivity() {
            return voiceActivity;
        }

        public void setVoiceActivity(boolean voiceActivity) {
            this.voiceActivity = voiceActivity;
        }

        public long getCheckSeq() {
            return checkSeq;
        }

        public void setCheckSeq(long checkSeq) {
            this.checkSeq = checkSeq;
        }
    }
}

